cilium-health/launch/endpoint.go

```diff
@@ -0,0 +1,211 @@
+		}
+	}()
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+	go func() {
+		in := bufio.NewScanner(stderr)
+		for in.Scan() {
+			scopedLog.Infof("%s", in.Text())
+		}
+	}()
+
+	return nil
+}
+
+func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing) error {
+	routes := []plugins.Route{}
+	v4Routes, err := plugins.IPv4Routes(addressing)
+	if err == nil {
+		routes = append(routes, v4Routes...)
+	} else {
+		log.Debugf("Couldn't get IPv4 routes for health routing")
+	}
+	v6Routes, err := plugins.IPv6Routes(addressing)
+	if err != nil {
+		return fmt.Errorf("Failed to get IPv6 routes")
+	}
+	routes = append(routes, v6Routes...)
+
+	prog := "ip"
+	args := []string{"netns", "exec", netns, "bash", "-c"}
+	buf := bytes.Buffer{}
+	join := "&&"
+	for _, rt := range routes {
+		cmd := strings.Join(rt.ToIPCommand(dev), " ")
+		log.WithField("netns", netns).WithField("command", cmd).Info("Adding route")
+		buf.WriteString(cmd)
+		buf.WriteString(join)
+	}
+	buf.Truncate(buf.Len() - len(join))
+	args = append(args, buf.String())
+
+	log.Infof("Running \"%s %+v\"", prog, args)
+	out, err := exec.Command(prog, args...).CombinedOutput()
+	if err == nil {
+		log.Warn(out)
+	}
+
+	return err
+}
+
+// LaunchAsEndpoint launches the cilium-health agent in a nested network
+// namespace and attaches it to Cilium the same way as any other endpoint,
+// but with special reserved labels.
+func LaunchAsEndpoint(owner endpoint.Owner, hostAddressing *models.NodeAddressing, opts *option.BoolOptions) context.CancelFunc {
+	ip4, ip6, err := ipam.AllocateNext("")
+	if err != nil {
+		log.WithError(err).Fatal("Error while allocating cilium-health IP")
+	}
+
+	// XXX: Make the above IPs known to pkg/node
+
+	// XXX: Release IPs later on?
+
+	// Prepare the endpoint change request
+	id := int64(addressing.CiliumIPv6(ip6).EndpointID())
+	info := &models.EndpointChangeRequest{
+		ID:            id,
+		ContainerID:   endpoint.NewCiliumID(id),
+		ContainerName: "cilium-health",
+		State:         models.EndpointStateWaitingForIdentity,
+		Addressing: &models.EndpointAddressing{
+			IPV6: ip6.String(),
+			IPV4: ip4.String(),
+		},
+	}
+
+	// Create the veth pair for the health endpoint
+	vethName := "cilium_health"
+	vethPeerName := "cilium"
+	if link, err := netlink.LinkByName(vethName); err == nil {
+		netlink.LinkDel(link)
+	} else {
+		log.Debug("Didn't find existing link")
+	}
+	veth, _, err := plugins.SetupVethWithNames(vethName, vethPeerName, 1450, info)
+	if err != nil {
+		log.WithError(err).Fatal("Error while creating cilium-health veth")
+	}
+	defer func() {
+		if err == nil {
+			return
+		}
+		if err = netlink.LinkDel(veth); err != nil {
+			log.WithError(err).WithField(logfields.Veth, veth.Name).Warn("failed to clean up veth")
+		}
+	}()
+
+	// Invoke spawn_netns.sh to spin up the health endpoint
+	pidfile := fmt.Sprintf("%s/health-endpoint.pid", owner.GetStateDir())
+	os.RemoveAll(pidfile)
```

Comment:
Out of curiosity, shouldn't we prevent from starting this cilium-health instance if there's one already running? If yes, why are we deleting the old `pidfile`?

Reply:
We won't be able to get logs from the old one. I actually don't expect this condition to be hit, but arguably we should actually read the file if it exists and terminate the process. I'll look into this.

Reply:
I updated this code to open the file and terminate the process.